### DISCUSSION TITLE
Circular import errors with PyInstaller .exe files

### DISCUSSION THREAD

**Discussion by AmirV97** [Project-MONAI/MONAI#7758] [Q&A]
I have written a simple inference pipeline along with a very simple tkinter GUI. The script runs fine when i run it with a Python terminal. 
However, when I use PyInstaller to get a standalone exe file, i get circular import errors. I have already tried the following:
1. Fresh virtual env with only the necessary packages
2. Import statement as granular as possible
3. A dummy torch model with inference GUI and packaged using PyInstaller; it ran fine. From the error message i believe MONAI is the cause of the problem:
> File "monai\networks\layers\__init__.py", line 16, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
  File "monai\networks\layers\factories.py", line 70, in <module>
ImportError: cannot import name 'has_nvfuser_instance_norm' from partially initialized module 'monai.networks.utils' (most likely due to a circular import) (C:\Users\Amir\AppData\Local\Temp\_MEI51162\monai\networks\utils.pyc)

> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "test__.py", line 7, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\__init__.py", line 58, in <module>
>   File "monai\utils\module.py", line 221, in load_submodules
>   File "monai\utils\module.py", line 211, in load_submodules
>   File "importlib\__init__.py", line 127, in import_module
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\networks\__init__.py", line 14, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\networks\utils.py", line 30, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\apps\__init__.py", line 14, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\apps\datasets.py", line 32, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\data\__init__.py", line 29, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\data\dataset.py", line 39, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\transforms\__init__.py", line 16, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\transforms\croppad\array.py", line 31, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\transforms\croppad\functional.py", line 28, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\transforms\utils.py", line 29, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\networks\layers\__init__.py", line 16, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 419, in exec_module
>   File "monai\networks\layers\factories.py", line 70, in <module>
> ImportError: cannot import name 'has_nvfuser_instance_norm' from partially initialized module 'monai.networks.utils' (most likely due to a circular import) (C:\Users\Amir\AppData\Local\Temp\_MEI51162\monai\networks\utils.pyc)
> 
> Multiple versions of MONAI may have been installed?
> Please see the installation guide: https://docs.monai.io/en/stable/installation.html

here's my code for context:
```

from torch import nn
from torch.optim.swa_utils import AveragedModel
import torch
import tkinter as tk
from tkinter import filedialog
from monai.transforms.compose import Compose
from monai.transforms.post.array import Activations
from monai.transforms.post.array import AsDiscrete
from monai.transforms.io.dictionary import LoadImaged
from monai.transforms.transform import MapTransform
from monai.transforms.intensity.dictionary import NormalizeIntensityd
from monai.transforms.spatial.dictionary import Resized
from monai.transforms.croppad.dictionary import CropForegroundd
from monai.transforms.post.array import KeepLargestConnectedComponent
from monai.transforms.post.array import FillHoles
from monai.transforms.io.array import SaveImage

[model code]

preprocessing = Compose([
    LoadImaged(keys=["image"]),
    AddChannelDim(keys=['image']),
    NormalizeIntensityd(keys="image", nonzero=False, channel_wise=False, subtrahend=8.6037, divisor=7.5377),
    CropForegroundd(keys=['image'], source_key='image'),
    Resized(keys=['image'], spatial_size=(160, 160, 160), mode=['trilinear']),
])

post_trans = Compose([
    Activations(sigmoid=True),
    AsDiscrete(threshold=0.5),
    KeepLargestConnectedComponent(connectivity=2),
    FillHoles(connectivity=2),
    ])

saver = SaveImage()
def inference(img_dir):
    inp = {'image':str(img_dir)}
    inp = preprocessing(inp)
    model.eval()
    with torch.inference_mode():
        out = model(inp['image'].unsqueeze(dim=0))
    out = post_trans(out).squeeze()

    saver(out)

def run_inference():
    img_dir = entry.get()
    result = inference(img_dir)
    result_label.config(text=result)

def browse_file():
    file_path = filedialog.askopenfilename(filetypes=[("NII Files", "*.nii")])
    entry.delete(0, tk.END)
    entry.insert(tk.END, file_path)

window = tk.Tk()
entry_label = tk.Label(window, text="Enter .nii file directory:")
entry_label.pack()
entry = tk.Entry(window)
entry.pack()
browse_button = tk.Button(window, text="Browse", command=browse_file)
browse_button.pack()
infer_button = tk.Button(window, text="Run Inference", command=run_inference)
infer_button.pack()
result_label = tk.Label(window, text="")
result_label.pack()
window.mainloop()
```

**Comment by KumoLiu** [comment]
Hi @AmirV97, perhaps this https://github.com/Project-MONAI/MONAI/discussions/4197 can help you.

Hope it can help you, thanks!

**Reply by AmirV97** [reply]
The mentioned code snippet uses the class Path in line 2, which is not imported. I'm using a .py script to get my exe file:
import PyInstaller.main
PyInstaller.main.run([
'test__.spec',
])

test__.spec contain the code snippet. I'm getting the following error:
NameError: name 'Path' is not defined

I've added the line from pathlib import Path to both test__.spec and the .py script but i'm still getting the error.

Considering there have been many instances of this problem posted here (and all of them being told to check said comment), a more comprehensive answer would be much appreciated.

Thanks!
